dotbot/protocol.py

```diff
@@ -59,543 +62,485 @@ class ProtocolPayloadParserException(Exception):
-    version: int = PROTOCOL_VERSION
-    msg_id: int = 0
 
-    @property
-    def fields(self) -> List[ProtocolField]:
-        return [
-            ProtocolField(self.destination, name="dst", length=8),
-            ProtocolField(self.source, name="src", length=8),
-            ProtocolField(self.swarm_id, name="swarm id", length=2),
-            ProtocolField(self.application, name="app."),
-            ProtocolField(self.version, name="ver."),
-            ProtocolField(self.msg_id, name="msg id", length=4),
+
+@dataclass
+class PayloadAdvertisement(Packet):
+    """Dataclass that holds an advertisement (emtpy)."""
+
+    metadata: list[PacketFieldMetadata] = dataclasses.field(
+        default_factory=lambda: [
+            PacketFieldMetadata(name="application", disp="app"),
         ]
+    )
 
-    @staticmethod
-    def from_bytes(bytes_) -> ProtocolData:
-        return ProtocolHeader(
-            int.from_bytes(bytes_[0:8], "little"),  # destination
-            int.from_bytes(bytes_[8:16], "little"),  # source
-            int.from_bytes(bytes_[16:18], "little"),  # swarm_id
-            ApplicationType(int.from_bytes(bytes_[18:19], "little")),  # application
-            int.from_bytes(bytes_[19:20], "little"),  # version
-            int.from_bytes(bytes_[20:24], "little"),  # msg id
-        )
+    application: ApplicationType = ApplicationType.DotBot
 
 
 @dataclass
-class CommandMoveRaw(ProtocolData):
+class PayloadCommandMoveRaw(Packet):
     """Dataclass that holds move raw command data fields."""
 
+    metadata: list[PacketFieldMetadata] = dataclasses.field(
+        default_factory=lambda: [
+            PacketFieldMetadata(name="lelf_x", disp="lx", signed=True),
+            PacketFieldMetadata(name="lelf_y", disp="ly", signed=True),
+            PacketFieldMetadata(name="right_y", disp="rx", signed=True),
+            PacketFieldMetadata(name="right_y", disp="ry", signed=True),
+        ]
+    )
+
     left_x: int = 0
     left_y: int = 0
     right_x: int = 0
     right_y: int = 0
 
-    @property
-    def fields(self) -> List[ProtocolField]:
-        return [
-            ProtocolField(self.left_x, name="lx", length=1, signed=True),
-            ProtocolField(self.left_y, name="ly", length=1, signed=True),
-            ProtocolField(self.right_x, name="rx", length=1, signed=True),
-            ProtocolField(self.right_y, name="ry", length=1, signed=True),
-        ]
-
-    @staticmethod
-    def from_bytes(bytes_) -> ProtocolData:
-        return CommandMoveRaw(*bytes_[0:4])
-
 
 @dataclass
-class CommandRgbLed(ProtocolData):
+class PayloadCommandRgbLed(Packet):
     """Dataclass that holds a complete rgb led command fields."""
 
+    metadata: list[PacketFieldMetadata] = dataclasses.field(
+        default_factory=lambda: [
+            PacketFieldMetadata(name="red", disp="red"),
+            PacketFieldMetadata(name="green", disp="green"),
+            PacketFieldMetadata(name="blue", disp="blue"),
```

Comment:
Could make `disp` optional and then add a `__post_init__` to `PacketFieldMetadata` that copies `name` into `disp` in case `disp` is empty.